<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.17"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioExplorer: Class Hierarchy</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="BBBE_icon_64.png"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">BioExplorer
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.17 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('hierarchy.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">Class Hierarchy</div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock">
<p><a href="inherits.html">Go to the graphical class hierarchy</a></p>
This inheritance list is sorted roughly, but not completely, alphabetically:</div><div class="directory">
<div class="levels">[detail level <span onclick="javascript:toggleLevel(1);">1</span><span onclick="javascript:toggleLevel(2);">2</span><span onclick="javascript:toggleLevel(3);">3</span>]</div><table class="directory">
<tr id="row_0_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_0_" class="arrow" onclick="toggleFolder('0_')">&#9660;</span><span class="icona"><span class="icon">C</span></span><b>AbstractSimulationHandler</b></td><td class="desc"></td></tr>
<tr id="row_0_0_"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="dc/d44/classbioexplorer_1_1FieldsHandler.html" target="_self">bioexplorer::FieldsHandler</a></td><td class="desc">Handles electro-magnetic fields data structures </td></tr>
<tr id="row_1_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d6/d27/structbioexplorer_1_1AddGridDetails.html" target="_self">bioexplorer::AddGridDetails</a></td><td class="desc">Defines the parameters needed when adding 3D grid in the scene </td></tr>
<tr id="row_2_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d9/db8/structbioexplorer_1_1AminoAcid.html" target="_self">bioexplorer::AminoAcid</a></td><td class="desc">Structure containing amino acids long and shot names </td></tr>
<tr id="row_3_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d2/de1/structbioexplorer_1_1AminoAcidDetails.html" target="_self">bioexplorer::AminoAcidDetails</a></td><td class="desc">Structure used to set an amino acid in protein sequences </td></tr>
<tr id="row_4_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d3/dac/structbioexplorer_1_1AminoAcidInformationDetails.html" target="_self">bioexplorer::AminoAcidInformationDetails</a></td><td class="desc"></td></tr>
<tr id="row_5_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d7/d35/structbioexplorer_1_1AminoAcidSequenceAsRangesDetails.html" target="_self">bioexplorer::AminoAcidSequenceAsRangesDetails</a></td><td class="desc">Structure defining a selection of amino acids on a protein of an assembly. The selection is defined as a range of indices </td></tr>
<tr id="row_6_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d2/d03/structbioexplorer_1_1AminoAcidSequenceAsStringDetails.html" target="_self">bioexplorer::AminoAcidSequenceAsStringDetails</a></td><td class="desc">Structure defining a selection of amino acids on a protein of an assembly. The selection is defined as a string </td></tr>
<tr id="row_7_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="db/db7/classbioexplorer_1_1Assembly.html" target="_self">bioexplorer::Assembly</a></td><td class="desc">Container for biological entities (proteins, membranes, sugars, etc.) </td></tr>
<tr id="row_8_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="dd/d75/structbioexplorer_1_1AssemblyDetails.html" target="_self">bioexplorer::AssemblyDetails</a></td><td class="desc"><a class="el" href="db/db7/classbioexplorer_1_1Assembly.html" title="The Assembly class is a container for biological entities (proteins, membranes, sugars,...">Assembly</a> representation </td></tr>
<tr id="row_9_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d3/d4a/structbioexplorer_1_1AssemblyTransformationsDetails.html" target="_self">bioexplorer::AssemblyTransformationsDetails</a></td><td class="desc">Structure defining transformations to apply to assembly elements </td></tr>
<tr id="row_10_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="db/d01/structbioexplorer_1_1Atom.html" target="_self">bioexplorer::Atom</a></td><td class="desc">Structure containing information about an atom, as stored in a PDB file </td></tr>
<tr id="row_11_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="dd/d9e/structbioexplorer_1_1BuildFieldsDetails.html" target="_self">bioexplorer::BuildFieldsDetails</a></td><td class="desc">Structure containing information about how to build magnetic fields from atom positions and charge </td></tr>
<tr id="row_12_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="de/d84/structbioexplorer_1_1BuildPointCloudDetails.html" target="_self">bioexplorer::BuildPointCloudDetails</a></td><td class="desc">Structure defining how to build a point cloud from the scene </td></tr>
<tr id="row_13_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_13_" class="arrow" onclick="toggleFolder('13_')">&#9660;</span><span class="icona"><span class="icon">C</span></span><b>Camera</b></td><td class="desc"></td></tr>
<tr id="row_13_0_"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d3/d23/structbioexplorer_1_1PerspectiveStereoCamera.html" target="_self">bioexplorer::PerspectiveStereoCamera</a></td><td class="desc"></td></tr>
<tr id="row_14_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d7/d0c/structbioexplorer_1_1ColorSchemeDetails.html" target="_self">bioexplorer::ColorSchemeDetails</a></td><td class="desc">Defines the color scheme to apply to a protein </td></tr>
<tr id="row_15_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d8/d69/structbioexplorer_1_1DatabaseAccessDetails.html" target="_self">bioexplorer::DatabaseAccessDetails</a></td><td class="desc">Structure defining how to export data into a DB </td></tr>
<tr id="row_16_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d3/d1a/classbioexplorer_1_1DBConnector.html" target="_self">bioexplorer::DBConnector</a></td><td class="desc">Allows the BioExplorer to communicate with a PostgreSQL database. The <a class="el" href="d3/d1a/classbioexplorer_1_1DBConnector.html" title="The DBConnector class allows the BioExplorer to communicate with a PostgreSQL database....">DBConnector</a> requires the pqxx library to be found at compilation time </td></tr>
<tr id="row_17_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_17_" class="arrow" onclick="toggleFolder('17_')">&#9660;</span><span class="icona"><span class="icon">C</span></span><b>ExtensionPlugin</b></td><td class="desc"></td></tr>
<tr id="row_17_0_" class="even"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d7/d7b/classbioexplorer_1_1BioExplorerPlugin.html" target="_self">bioexplorer::BioExplorerPlugin</a></td><td class="desc">This class implements the BioExplorer plugin </td></tr>
<tr id="row_18_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="db/daa/structbioexplorer_1_1FileAccessDetails.html" target="_self">bioexplorer::FileAccessDetails</a></td><td class="desc">Structure defining how to export data into a file </td></tr>
<tr id="row_19_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d4/ddc/classbioexplorer_1_1GeneralSettings.html" target="_self">bioexplorer::GeneralSettings</a></td><td class="desc"><a class="el" href="d4/ddc/classbioexplorer_1_1GeneralSettings.html" title="GeneralSettings is a singleton class that holds general settings for the plugin.">GeneralSettings</a> is a singleton class that holds general settings for the plugin </td></tr>
<tr id="row_20_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="da/d8a/structbioexplorer_1_1GeneralSettingsDetails.html" target="_self">bioexplorer::GeneralSettingsDetails</a></td><td class="desc">Structure defining the plugin general settings </td></tr>
<tr id="row_21_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_21_" class="arrow" onclick="toggleFolder('21_')">&#9660;</span><span class="icona"><span class="icon">C</span></span><b>Loader</b></td><td class="desc"></td></tr>
<tr id="row_21_0_"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d9/d08/classbioexplorer_1_1CacheLoader.html" target="_self">bioexplorer::CacheLoader</a></td><td class="desc"></td></tr>
<tr id="row_22_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_22_" class="arrow" onclick="toggleFolder('22_')">&#9660;</span><span class="icona"><span class="icon">C</span></span><b>Material</b></td><td class="desc"></td></tr>
<tr id="row_22_0_"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="da/db3/structbioexplorer_1_1AdvancedMaterial.html" target="_self">bioexplorer::AdvancedMaterial</a></td><td class="desc"></td></tr>
<tr id="row_23_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d7/dff/structbioexplorer_1_1MaterialIdsDetails.html" target="_self">bioexplorer::MaterialIdsDetails</a></td><td class="desc">List of material identifiers attached to a Brayns model </td></tr>
<tr id="row_24_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d6/df4/structbioexplorer_1_1MaterialsDetails.html" target="_self">bioexplorer::MaterialsDetails</a></td><td class="desc">Structure containing attributes of materials attached to one or several Brayns models </td></tr>
<tr id="row_25_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d9/df5/structbioexplorer_1_1MembraneDetails.html" target="_self">bioexplorer::MembraneDetails</a></td><td class="desc">A <a class="el" href="d6/dca/classbioexplorer_1_1Membrane.html" title="A Membrane object implements a 3D structure of a given shape, but with a surface composed of instance...">Membrane</a> is a shaped assembly of phospholipids </td></tr>
<tr id="row_26_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d0/d14/structbioexplorer_1_1MeshBasedMembraneDetails.html" target="_self">bioexplorer::MeshBasedMembraneDetails</a></td><td class="desc">Data structure describing a membrane based on the shape of a mesh </td></tr>
<tr id="row_27_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d9/d55/structbioexplorer_1_1ModelIdDetails.html" target="_self">bioexplorer::ModelIdDetails</a></td><td class="desc">Model identifier </td></tr>
<tr id="row_28_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d6/d54/structbioexplorer_1_1ModelIdFileAccessDetails.html" target="_self">bioexplorer::ModelIdFileAccessDetails</a></td><td class="desc"></td></tr>
<tr id="row_29_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d4/d8b/structbioexplorer_1_1ModelsVisibilityDetails.html" target="_self">bioexplorer::ModelsVisibilityDetails</a></td><td class="desc">Structure defining how visible models are in the scene </td></tr>
<tr id="row_30_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_30_" class="arrow" onclick="toggleFolder('30_')">&#9660;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="da/dba/classbioexplorer_1_1Node.html" target="_self">bioexplorer::Node</a></td><td class="desc">The <a class="el" href="da/dba/classbioexplorer_1_1Node.html" title="The Node class.">Node</a> class </td></tr>
<tr id="row_30_0_" class="even"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d6/dca/classbioexplorer_1_1Membrane.html" target="_self">bioexplorer::Membrane</a></td><td class="desc">A <a class="el" href="d6/dca/classbioexplorer_1_1Membrane.html" title="A Membrane object implements a 3D structure of a given shape, but with a surface composed of instance...">Membrane</a> object implements a 3D structure of a given shape, but with a surface composed of instances of one or several proteins </td></tr>
<tr id="row_30_1_"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="dd/d42/classbioexplorer_1_1MeshBasedMembrane.html" target="_self">bioexplorer::MeshBasedMembrane</a></td><td class="desc">A <a class="el" href="dd/d42/classbioexplorer_1_1MeshBasedMembrane.html" title="A MeshBasedMembrane object implements a 3D structure that has the shape of a given OBJ Mesh,...">MeshBasedMembrane</a> object implements a 3D structure that has the shape of a given OBJ Mesh, but with a surface composed of instances of a protein. This class is typicaly used to create membranes with a shape provided by a 3D Mesh </td></tr>
<tr id="row_30_2_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_30_2_" class="arrow" onclick="toggleFolder('30_2_')">&#9660;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d6/d81/classbioexplorer_1_1Molecule.html" target="_self">bioexplorer::Molecule</a></td><td class="desc">Implements the 3D representation of a molecule. The object also contains metadata attached to the molecule itself, such as the amino acids sequence, or the chain ids for example. The current implementation only supports PDB as an input format for the molecule data and metadata </td></tr>
<tr id="row_30_2_0_"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="df/dc3/classbioexplorer_1_1Glycans.html" target="_self">bioexplorer::Glycans</a></td><td class="desc">The <a class="el" href="df/dc3/classbioexplorer_1_1Glycans.html" title="The Glycans class.">Glycans</a> class </td></tr>
<tr id="row_30_2_1_" class="even"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="db/d87/classbioexplorer_1_1Protein.html" target="_self">bioexplorer::Protein</a></td><td class="desc">The <a class="el" href="db/d87/classbioexplorer_1_1Protein.html" title="The Protein class.">Protein</a> class </td></tr>
<tr id="row_30_3_"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d2/d77/classbioexplorer_1_1RNASequence.html" target="_self">bioexplorer::RNASequence</a></td><td class="desc">The <a class="el" href="d2/d77/classbioexplorer_1_1RNASequence.html" title="The RNASequence class.">RNASequence</a> class </td></tr>
<tr id="row_31_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="df/d61/structbioexplorer_1_1Nucleotid.html" target="_self">bioexplorer::Nucleotid</a></td><td class="desc">Structure representing a nucleotid </td></tr>
<tr id="row_32_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d1/d8e/classbioexplorer_1_1Octree.html" target="_self">bioexplorer::Octree</a></td><td class="desc">Implements the <a class="el" href="d1/d8e/classbioexplorer_1_1Octree.html" title="The Octree class implements the Octree acceleration structure used by the FieldsRenderer class to ren...">Octree</a> acceleration structure used by the <a class="el" href="d0/d37/classbioexplorer_1_1FieldsRenderer.html" title="The FieldsRenderer class allows visualization of magnetic fields created by atoms in the 3D scene....">FieldsRenderer</a> class to render magnetic fields </td></tr>
<tr id="row_33_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d9/dfb/classbioexplorer_1_1OctreeNode.html" target="_self">bioexplorer::OctreeNode</a></td><td class="desc">Implement a spherical node of the <a class="el" href="d1/d8e/classbioexplorer_1_1Octree.html" title="The Octree class implements the Octree acceleration structure used by the FieldsRenderer class to ren...">Octree</a> acceleration structure used by the Fields renderer </td></tr>
<tr id="row_34_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d0/d4b/classbioexplorer_1_1OOCManager.html" target="_self">bioexplorer::OOCManager</a></td><td class="desc">The <a class="el" href="d0/d4b/classbioexplorer_1_1OOCManager.html" title="The OOCManager classes manager the out-of-core engine of the BioExplorer. The scene is devided into b...">OOCManager</a> classes manager the out-of-core engine of the BioExplorer. The scene is devided into bricks of a given size that are, in the current implementation, stored in a PostgreSQL database. The out-of-core engine is activated via command line parameters only. By settings the &ndash;ooc-enabled command line parameter, the out-of-core engine is enabled and, in order to avoid unnecessary network traffic, the scene contents are not broadcasted anymore to the BioExplorer clients (Python notebook or User Inferface). Note that ENV_ROCKETS_DISABLE_SCENE_BROADCASTING environment variable has to be set to order Brayns not to broadcast the scene changes. The out-of-core mode is a read-only feature and does not allow any modification of the scene. When the camera position changes, the <a class="el" href="d0/d4b/classbioexplorer_1_1OOCManager.html" title="The OOCManager classes manager the out-of-core engine of the BioExplorer. The scene is devided into b...">OOCManager</a> identifies the visible bricks and loads them as a background task. For performance reasons, models are loaded as invisible. This allows the ray-tracing engine to create the BVH before the model is added to the scene. Once all models are loaded, they are added to the scene. Note that the <a class="el" href="d0/d4b/classbioexplorer_1_1OOCManager.html" title="The OOCManager classes manager the out-of-core engine of the BioExplorer. The scene is devided into b...">OOCManager</a> only does work when the frame buffer accumulation is greater than 1, which means that the camera is not moving. If the &ndash;ooc-unload-bricks command line parameter is set, invisible bricks are unloaded from memory and removed from the scene. The &ndash;ooc-update-frequency command line parameters defines (in seconds) at which frequency models should be loaded. The &ndash;ooc-nb-bricks-per-cycle command line parameter defines how many bricks can be loaded at each rendering cycle. The connection to the database is defined by the &ndash;ooc-db-host, &ndash;ooc-db-port, &ndash;ooc-db-name,&ndash;ooc-db-user,&ndash;ooc-db-password and &ndash;ooc-db-schema parameters. If the BioExplorer fails to find the pqxx dependency to connect to the PostgreSQL database, it falls back to a file-based mode where bricks can be stored as BioExplorer cache files, in the folder defined by the &ndash;ooc-bricks-folder command line argument. The &ndash;ooc-show-grid command line parameter can be used to show a grid corresponding the positions of the bricks in the scene </td></tr>
<tr id="row_35_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d8/db3/structbioexplorer_1_1OOCSceneConfigurationDetails.html" target="_self">bioexplorer::OOCSceneConfigurationDetails</a></td><td class="desc"></td></tr>
<tr id="row_36_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d0/d39/classbioexplorer_1_1PointCloudMesher.html" target="_self">bioexplorer::PointCloudMesher</a></td><td class="desc"></td></tr>
<tr id="row_37_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="da/df7/structbioexplorer_1_1ProteinDetails.html" target="_self">bioexplorer::ProteinDetails</a></td><td class="desc"></td></tr>
<tr id="row_38_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="db/d82/structbioexplorer_1_1ProteinInstanceTransformationDetails.html" target="_self">bioexplorer::ProteinInstanceTransformationDetails</a></td><td class="desc"></td></tr>
<tr id="row_39_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="df/db9/structbioexplorer_1_1RandomizationInformation.html" target="_self">bioexplorer::RandomizationInformation</a></td><td class="desc"></td></tr>
<tr id="row_40_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_40_" class="arrow" onclick="toggleFolder('40_')">&#9660;</span><span class="icona"><span class="icon">C</span></span><b>Renderer</b></td><td class="desc"></td></tr>
<tr id="row_40_0_" class="even"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="de/d08/classbioexplorer_1_1AdvancedRenderer.html" target="_self">bioexplorer::AdvancedRenderer</a></td><td class="desc">Renderer that can perform global illumination (light shading, shadows, ambient occlusion, color bleeding, light emission) </td></tr>
<tr id="row_40_1_"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d0/d37/classbioexplorer_1_1FieldsRenderer.html" target="_self">bioexplorer::FieldsRenderer</a></td><td class="desc">Allows visualization of magnetic fields created by atoms in the 3D scene. An <a class="el" href="d1/d8e/classbioexplorer_1_1Octree.html" title="The Octree class implements the Octree acceleration structure used by the FieldsRenderer class to ren...">Octree</a> acceleration structure has to be built by the be_build_fields API in order to feed the renderer with the information needed to compute the value of the field for every point in the 3D space </td></tr>
<tr id="row_41_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d9/d72/structbioexplorer_1_1Response.html" target="_self">bioexplorer::Response</a></td><td class="desc"></td></tr>
<tr id="row_42_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="da/d3c/structbioexplorer_1_1RGBColor.html" target="_self">bioexplorer::RGBColor</a></td><td class="desc">Structure defining the RGB color of atoms according to the JMol Scheme </td></tr>
<tr id="row_43_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="dc/d0c/structbioexplorer_1_1RNASequenceDetails.html" target="_self">bioexplorer::RNASequenceDetails</a></td><td class="desc">RNA sequence descriptor </td></tr>
<tr id="row_44_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d9/db8/structbioexplorer_1_1Sequence.html" target="_self">bioexplorer::Sequence</a></td><td class="desc"><a class="el" href="d9/db8/structbioexplorer_1_1Sequence.html" title="Sequence of amino acids.">Sequence</a> of amino acids </td></tr>
<tr id="row_45_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d9/dab/structbioexplorer_1_1SugarsDetails.html" target="_self">bioexplorer::SugarsDetails</a></td><td class="desc">Data structure describing the glycans </td></tr>
<tr id="row_46_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="db/d99/classbioexplorer_1_1SurfaceMesher.html" target="_self">bioexplorer::SurfaceMesher</a></td><td class="desc"></td></tr>
<tr id="row_47_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="d4/dbf/classbioexplorer_1_1UniqueId.html" target="_self">bioexplorer::UniqueId</a></td><td class="desc">The UniqueID class provides a way to get a unique identifier accross the application </td></tr>
</table>
</div><!-- directory -->
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.17 </li>
  </ul>
</div>
</body>
</html>
