<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Blue Brain BioExplorer: Main Page</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="BBBE_icon_64.png"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Blue Brain BioExplorer
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('index.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="PageDoc"><div class="header">
  <div class="headertitle">
<div class="title">Blue Brain BioExplorer Documentation</div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="md_README"></a></p>
<h1><a class="anchor" id="autotoc_md1"></a>
Blue Brain BioExplorer</h1>
<table border="0">
<tr>
<td><a href="https://github.com/BlueBrain/BioExplorer/tags"><img src="https://img.shields.io/github/v/tag/BlueBrain/BioExplorer?style=for-the-badge" alt="Latest release" class="inline"/> </a>  </td><td><a href="https://github.com/BlueBrain/BioExplorer/blob/master/LICENSE.md"><img src="https://img.shields.io/github/license/BlueBrain/BioExplorer?style=for-the-badge" alt="License" class="inline"/> </a>  </td><td><a href="https://github.com/BlueBrain/BioExplorer/forks"><img src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Factions-badge.atrox.dev%2Fbluebrain%2Fbioexplorer%2Fbadge%3Fref%3Dmaster&amp;style=for-the-badge" alt="Actions" style="pointer-events: none;" class="inline"/> </a>  </td><td><a href="https://github.com/BlueBrain/BioExplorer/stargazers"><img src="https://img.shields.io/github/stars/BlueBrain/BioExplorer?style=for-the-badge" alt="Stars" class="inline"/> </a>  </td><td><a href="https://github.com/BlueBrain/BioExplorer/network/members"><img src="https://img.shields.io/github/forks/BlueBrain/BioExplorer?style=for-the-badge" alt="Forks" class="inline"/> </a>  </td><td><a href="http://www.pydocstyle.org/"><img src="https://img.shields.io/badge/docstrings-pydocstyle-informational?style=for-the-badge" alt="Pydocstyle" class="inline"/> </a>  </td><td><a href="https://pypi.org/project/pycodestyle/"><img src="https://img.shields.io/badge/docstrings-pycodestyle-informational?style=for-the-badge" alt="Pycodestyle" class="inline"/> </a>   </td></tr>
</table>
<p><img src="bioexplorer_banner.png" alt="" class="inline" title="___"/></p>
<h2><a class="anchor" id="autotoc_md2"></a>
Description</h2>
<p>Exploration relies on building software that combines data integration, analysis and interactive visualization to build, modify and navigate through large scientific datasets. For this, Blue Brain built and open-sourced the Blue Brain BioExplorer. The Blue Brain BioExplorer (BBBE), which started as an internal project, is now open source. It was originally developed to answer key scientific questions related to the Coronavirus as a use case and to deliver a visualization tool. Today, the <em>BBBE</em> allows to reconstruct, visualize, explore and describe in detail the structure and function of highly-detailed biological structures such as molecular systems, neurons, astrocytes, blood vessels, and more. You can see the first application of the <em>BBBE</em> in '<a href="https://www.frontiersin.org/articles/10.3389/fpubh.2021.695139/full?utm_source=fweb&amp;utm_medium=nblog&amp;utm_campaign=ba-sci-fpubh-covid-19-elevated-blood-glucose-blue-brain">A Machine-Generated View of the Role of Blood Glucose Levels in the Severity of COVID-19</a>' study, and see the SARS-COV-2 coronavirus as you have never seen it before by clicking on the following image!</p>
<div align="center"> <a href="https://youtu.be/Czn-YGLGfiA"><img src="https://github.com/BlueBrain/BioExplorer/blob/master/bioexplorer/pythonsdk/doc/source/images/covid19_movie_preview.png" alt="" style="width:50%;" class="inline"/> </a> </div><p>On the third anniversary of the first lock-down, we look back at how #MachineLearning helped reveal the role of blood glucose levels in the severity of COVID-19. Imagine what else could be tackled with OpenScience.</p>
<div align="center"> <a href="https://lnkd.in/e57vqgcq"><img src="https://github.com/BlueBrain/BioExplorer/blob/master/bioexplorer/pythonsdk/doc/source/images/covid19_documentary_preview.png" alt="" style="width:50%;" class="inline"/> </a> </div><h2><a class="anchor" id="autotoc_md3"></a>
The Harvard Brain</h2>
<p>Do you want want to know more about the full story? Read the <a href="http://www.theharvardbrain.com/spring-2023-8203lara-ota-buse-toksoumlz-and-kei-hayashi.html">Studies In Silico: An Interview With Cyrille Favreau On EPFL’s Blue Brain Project</a>.</p>
<h2><a class="anchor" id="autotoc_md4"></a>
Social Media</h2>
<ul>
<li><a href="https://www.linkedin.com/feed/update/urn:li:activity:7057015980253696000/">Blue Brain BioExplorer 1.6 goes RTX!</a></li>
<li><a href="https://www.linkedin.com/posts/blue-brain-project_machine-learning-and-understanding-the-role-activity-7044953592570535937-s6IB?utm_source=share&amp;utm_medium=member_desktop">On the third anniversary of the first lockdown, we look back at how Machine Learning helped reveal the role of blood glucose levels in the severity of COVID-19.</a></li>
<li><a href="https://www.linkedin.com/posts/blue-brain-project_if-youre-traveling-to-spain-over-the-holiday-activity-7009472054114533376-Serk?utm_source=share&amp;utm_medium=member_desktop">If you’re traveling to Spain over the holiday break, don’t miss the amazing exhibition Brain(s) opening soon in Madrid featuring images from the EPFL Blue Brain Project. Join the more than 86'000 visitors who already enjoyed it while on its Barcelona stop!</a></li>
<li><a href="https://www.linkedin.com/posts/blue-brain-project_openscience-neuroscience-scientificvisualization-activity-6980420229633867776-ViVr?utm_source=share&amp;utm_medium=member_desktop">The BlueBrainBioExplorer 1.4 is now out and ready to explore additional bio assemblies from neurons and astrocytes to vasculatures and more!</a></li>
<li><a href="https://www.linkedin.com/posts/blue-brain-project_cervellscccb-cerebroscccb-brainscccb-activity-6957702754731450369-gsLV?utm_source=share&amp;utm_medium=member_desktop">The EPFL Blue Brain Project will be featuring in the exciting Brain(s) exhibition at the Barcelona Centre of Contemporary Culture starting tomorrow!</a></li>
</ul>
<h2><a class="anchor" id="autotoc_md5"></a>
BioExplorer at PASC 2023</h2>
<div align="center"> <img src="https://www.pasc-conference.org/wp-content/uploads/2022/07/PASC_23_2.jpg" alt="Image" width="25%" height="25%" class="inline"/> </div><p> On the 27th of June 2023, the Blue Brain BioExplorer was presented at the PASC23 conference in Davos (Switzerland), as part of the <em>MS4E - Scientific Visualization of Big Data</em> minisymposium: <a href="https://pasc23.pasc-conference.org/presentation/?id=msa164&amp;sess=sess168">Visual Scientific Exploration at Blue Brain, and Beyond</a></p>
<h2><a class="anchor" id="autotoc_md6"></a>
BioExplorer goes RTX!</h2>
<p>BioExplorer now benefits from <a href="https://developer.nvidia.com/rtx/ray-tracing/optix">NVIDIA OptiX</a> backend compatibility, allowing it to fully utilize the NVIDIA RTX acceleration hardware capabilities for fast and high-quality rendering of scientific datasets. This advanced feature, along with the new stereo perspective camera, enables virtual reality use-cases with pure ray-tracing. Additionally, an AI-powered denoiser has been incorporated to enhance the real-time rendering quality.</p>
<p>Simply add the OptiX engine command line argument when starting the <em>BBBE</em>. </p><div class="fragment"><div class="line">service --engine optix6 --plugin BioExplorer</div>
</div><!-- fragment --><p>Note that, in order to offer the best experience on local desktops, the <em>BBBE</em> now also has a native viewer. </p><div class="fragment"><div class="line">viewer --engine optix6 --plugin BioExplorer</div>
</div><!-- fragment --><h2><a class="anchor" id="autotoc_md7"></a>
At the museum</h2>
<p>July 2022: The EPFL Blue Brain Project is featuring in the exciting <a href="https://www.cccb.org/en/exhibitions/file/brains/237851">Brain(s) exhibition</a> at the Barcelona Centre of Contemporary Culture at the Fundación Telefónica Madrid.</p>
<div align="center"> <img src="https://actu.epfl.ch/image/121700/1108x622.jpg" alt="Image" width="50%" height="50%" class="inline"/> </div><p>The Blue Brain images on show were created with the <em>BBBE</em>. Checkout the <a href="https://actu.epfl.ch/news/blue-brain-images-on-show-at-the-brains-exhibition/">news article</a> for more information.</p>
<h2><a class="anchor" id="autotoc_md8"></a>
Architecture</h2>
<p>The <em>BBBE</em> application is built on top of a fork of <a href="https://github.com/BlueBrain/Brayns/releases/tag/1.0.0">Blue Brain Brayns</a> 1.0.0 , the Blue Brain rendering platform. The <em>BBBE</em> uses the underlying technical capabilities of the rendering platform to create large scale and accurate 3D scenes from Jupyter notebooks.</p>
<p><img src="architecture.png" alt="" class="inline" title="___"/></p>
<p>More information can be found in the <a class="el" href="d1/dbc/md_ARCHITECTURE.html">architecture</a> documentation.</p>
<h2><a class="anchor" id="autotoc_md9"></a>
General components</h2>
<h3><a class="anchor" id="autotoc_md10"></a>
Assemblies</h3>
<p>Assemblies are groups of biological elements, such as proteins, membranes, glycans, etc. As an example, a virion is made of a lipid membrane, spikes proteins, an RNA sequence, etc, and all those elements belong to the same object. That’s why they need to belong to the same container, the assembly. Assemblies can have different shapes: <a class="el" href="d8/d0f/structSphere.html">Sphere</a>, Cube, etc, that are automatically generated according to the parameters of individual components.</p>
<h3><a class="anchor" id="autotoc_md11"></a>
Proteins</h3>
<p>Proteins are loaded from PDB files. Atoms, non-polymer chemicals and bonds can be loaded and displayed in various color schemes: chain id, atom, residue, etc. Proteins also contain the amino acid sequences of the individual chains. Sequences that can be used to query glycosylation sites, or functional regions of the protein.</p>
<p><img src="bioexplorer_proteins_banner.png" alt="" class="inline" title="___"/></p>
<h3><a class="anchor" id="autotoc_md12"></a>
Glycans</h3>
<p>Glycans are small proteins that are attached to an existing protein of the assembly. Individual glycan trees are loaded from PDB files and attached to the glycosylation sites of the specified protein. By default, glycans are attached to all available glycosylation sites, but a set of specific sites can be specified.</p>
<p>Glycan trees models located in the python sdk test folder were generated with <a href="http://glycam.org">Glycam Builder</a>.</p>
<p><img src="bioexplorer_glycans_banner.png" alt="" class="inline" title="___"/></p>
<h3><a class="anchor" id="autotoc_md13"></a>
RNA sequence</h3>
<p>An RNA sequence can be loaded from a text sequence of codons. Various shapes can be selected to represent the RNA sequence: Trefoil knot, torus, star, etc. This allows the sequence to be efficiently packed into a given volume. A different color is assigned per type of codon.</p>
<h3><a class="anchor" id="autotoc_md14"></a>
Mesh-based membranes</h3>
<p>Mesh-based membranes create membranes based on 3D meshes. This allows the construction of complex membranes where mesh faces are filled with proteins.</p>
<h3><a class="anchor" id="autotoc_md15"></a>
Virus</h3>
<p>A viral particle is an assembly consisting of a membrane, an RNA sequence, and a given number of S, M and E proteins. The virus has a predefined spherical shape defined by its radius. The default parameters for the virus are a radius of 45 nanometers, 62 S proteins, 42 E proteins, and 50 M proteins. Dimensions and concentrations were retrieved from the literature.</p>
<p><img src="bioexplorer_coronavirus_banner.png" alt="" class="inline" title="___"/></p>
<h3><a class="anchor" id="autotoc_md16"></a>
Membrane</h3>
<p>A membrane is an assembly of phospholipids. Phospholipids structures are created following the process described in the <a href="https://www.ks.uiuc.edu/Research/vmd">VMD</a> Membrane Proteins <a href="http://www.ks.uiuc.edu/Training/Tutorials">tutorial</a>. The assembly itself is generated by the BioExplorer, for a given shape, and a number of instances of phospholipids.</p>
<p><img src="bioexplorer_membrane_banner.png" alt="" class="inline" title="___"/></p>
<h3><a class="anchor" id="autotoc_md17"></a>
Vasculature</h3>
<p>Vasculatures are loaded from the database (see the database <a href="./storage/database/04_bioexplorer_vasculature_schema.sql">schema</a> and the example <a href="./bioexplorer/pythonsdk/notebooks/vasculature/BioExplorer_import_sonata_to_db.ipynb">notebook</a> for loading data from <a href="https://github.com/AllenInstitute/sonata">Sonata</a> files). A vasculature is defined by the blood vessels or arrangement of blood vessels in an organ or part.</p>
<div class="fragment"><div class="line">vasculature_model = bio_explorer.add_vasculature(</div>
<div class="line">    assembly_name=&#39;Vasculature&#39;, population_name=&#39;vasculature&#39;, use_sdf=True)</div>
</div><!-- fragment --><p><img src="bioexplorer_vasculature_banner.png" alt="" class="inline" title="___"/></p>
<p>An example dataset can be downloaded from the <a href="https://bbp.epfl.ch/ngv-portal/anatomy/experimental-data/">Blue Brain Neuro-Glia-Vasculature Portal</a></p>
<p>References:</p><ul>
<li><a href="https://www.cell.com/neuron/fulltext/S0896-6273(21)00080-5?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0896627321000805%3Fshowall%3Dtrue">Brain microvasculature has a common topology with local differences in geometry that match metabolic load</a></li>
<li><a href="https://bbp.epfl.ch/ngv-portal/#explore">Blue Brain Neuro-Glia-Vasculature Portal. Vasculature. Reconstruction Data</a></li>
</ul>
<h3><a class="anchor" id="autotoc_md18"></a>
Neurons</h3>
<p>Circuits of neurons are loaded from the database (see the database <a href="./storage/database/02_bioexplorer_neurons_schema.sql">schema</a> and the example <a href="./bioexplorer/pythonsdk/notebooks/neurons/BioExplorer_import_sonata_to_db.ipynb">notebook</a> for loading data from <a href="https://github.com/AllenInstitute/sonata">Sonata</a> files) using their position and orientation. Each cell is composed of sections that form the axons and dendrites, as well as spines. Cell internals such as the nucleus and the mitochondria can be automatically generated, according to the data provided by the scientific literature. The <em>BBBE</em> also implements procedural generation of varicosities along the axon, as well as myelin sheaths.</p>
<div class="fragment"><div class="line">status = bio_explorer.add_neurons(</div>
<div class="line">    assembly_name=&#39;Neurons&#39;,</div>
<div class="line">    population_name=&#39;neurons&#39;,</div>
<div class="line">    realism_level=be.MORPHOLOGY_REALISM_LEVEL_ALL,</div>
<div class="line">    synapses_type=be.NEURON_SYNAPSES_AFFERENT, generate_varicosities=True,</div>
<div class="line">    generate_internals=True, generate_externals=True</div>
<div class="line">)</div>
</div><!-- fragment --><p>An <a href="https://zenodo.org/record/6906785#.Ywym7tVBxH6">example dataset</a> is available in the <a href="https://www.biorxiv.org/content/10.1101/2022.08.11.503144v1">Model of Rat Non-barrel Somatosensory Cortex Anatomy</a> publication. Please refer to the documentation of the <a href="https://github.com/AllenInstitute/sonata">Sonata</a> format for information on how to load and analyze the model. A <a href="./bioexplorer/pythonsdk/notebooks/neurons/rat_non-barrel_somatosensory_cortex_anatomy/BioExplorer_rat_non-barrel_somatosensory_cortex_anatomy.ipynb">jupyter notebook</a> has been included with basic examples of how to explore the data using the Blue Brain open-source packages <a href="https://neurom.readthedocs.io/en/stable/">NeuroM</a> and <a href="https://bluebrainsnap.readthedocs.io/en/stable/">Blue Brain SNAP</a>.</p>
<p>References:</p><ul>
<li><a href="https://www.biorxiv.org/content/10.1101/2022.08.11.503144v1">A Model of Rat Non-barrel Somatosensory Cortex Anatomy</a></li>
<li><a href="https://www.sciencedirect.com/science/article/pii/S2451830118300293">A simplified morphological classification scheme for pyramidal cells in six layers of primary somatosensory cortex of juvenile rats</a></li>
<li><a href="https://doi.org/10.1002/cne.903610404">Age and sex do not affect the volume, cell numbers, or cell size of the suprachiasmatic nucleus of the rat: An unbiased stereological</a></li>
<li><a href="https://www.biorxiv.org/content/10.1101/2023.05.17.541167v1">Community-based Reconstruction and Simulation of a Full-scale Model of Region CA1 of Rat Hippocampus</a></li>
<li><a href="https://www.biorxiv.org/content/10.1101/2023.05.17.541168v2">Modeling and Simulation of Neocortical Micro- and Mesocircuitry. Part II: Physiology and Experimentation</a></li>
</ul>
<h3><a class="anchor" id="autotoc_md19"></a>
Astrocytes</h3>
<p>Circuits of astrocytes are loaded from the database (see the database <a href="./storage/database/03_bioexplorer_astrocytes_schema.sql">schema</a> and the example <a href="./bioexplorer/pythonsdk/notebooks/astrocytes/BioExplorer_import_sonata_to_db.ipynb">notebook</a> for loading data from <a href="https://github.com/AllenInstitute/sonata">Sonata</a> files) using their position and orientation. Astrocytes end-feet are connected to the vasculature using data stored in a dedicated connectome database <a href="./storage/database/05_bioexplorer_connectome_schema.sql">schema</a>. The <em>BBBE</em> allows end-feet to automatically adapt to the vasculature vessel size.</p>
<div class="fragment"><div class="line">vasculature_model = be.add_astrocytes(</div>
<div class="line">    assembly_name=&#39;Astrocytes&#39;, population_name=&#39;astrocytes&#39;, use_sdf=True)</div>
</div><!-- fragment --><p><img src="bioexplorer_neurons_banner.png" alt="" class="inline" title="___"/></p>
<p>The neurons assembly allows visualization of brain microcircuits. Morphologies imported from ASC, SWC or H5 files into the <em>BBBE</em> database can be loaded an transformed in different ways: simple spheres or somas only, simple rendering of full morphologies using spheres, cones and cylinders, or advanced rendering of full morphologies using the <a href="https://iquilezles.org/articles/distfunctions/">signed distance field</a> technique.</p>
<p>Multi-scale models of the rat and mouse brain integrate models of ion channels, single cells, microcircuits, brain regions, and brain systems at different levels of granularity (molecular models, morphologically detailed cellular models, and abstracted point neuron models). A neuronal microcircuit is the smallest functional ecosystem in any brain region that encompasses a diverse morphological and electrical assortment of neurons, and their synaptic interactions. Blue Brain has pioneered data-driven digital reconstructions and simulations of microcircuits to investigate how local neuronal structure gives rise to global network dynamics. These methods could be extended to digitally reconstruct microcircuits in any brain region.</p>
<p><img src="microcircuit.png" alt="" class="inline" title="___"/></p>
<p>In the nervous system, a synapse is a structure that permits a neuron (or nerve cell) to pass an electrical or chemical signal to another neuron. Synapses can be classified by the type of cellular structures serving as the pre- and post-synaptic components. The vast majority of synapses in the mammalian nervous system are classical axo-dendritic synapses (an axon connecting to a dendrite).</p>
<p><img src="synapses.png" alt="" class="inline" title="___"/></p>
<p>An example dataset is available on the <a href="https://bbp.epfl.ch/ngv-portal/anatomy/experimental-data/">Blue Brain Neuro-Glia-Vasculature Portal</a></p>
<p>References:</p><ul>
<li><a href="http://biorxiv.org/lookup/doi/10.1101/2022.02.28.482273">Thalamic control of sensory enhancement and sleep spindle properties in a biophysical model of thalamoreticular microcircuitry</a></li>
<li><a href="https://doi.org/10.1016/j.celrep.2022.110586">Computational synthesis of cortical dendritic morphologies</a></li>
<li><a href="https://doi.org/10.1093/cercor/bhab254">Digital reconstruction of the neuro-glia-vascular architecture</a></li>
<li><a href="https://bbp.epfl.ch/ngv-portal/anatomy/reconstruction-data/">Blue Brain Neuro-Glia-Vasculature Portal. Anatomy. Reconstruction Data</a></li>
<li><a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007696">The SONATA data format for efficient description of large-scale network models</a></li>
<li><a href="https://www.biorxiv.org/content/10.1101/2023.02.24.529863v1">Cortical cell assemblies and their underlying connectivity: an in silico study</a></li>
<li><a href="https://www.biorxiv.org/content/10.1101/2023.05.17.541168v1">Modeling and Simulation of Neocortical Micro- and Mesocircuitry. Part II: Physiology and Experimentation</a></li>
</ul>
<p><img src="synapses2.png" alt="" class="inline" title="___"/></p>
<h3><a class="anchor" id="autotoc_md20"></a>
Spinal cord</h3>
<p>The spinal cord is an elongated nervous structure that together with the brain forms the central nervous system. It relays sensory and motor information between the brain and the body, thus controlling most somatic and autonomic body functions. In recent years, great progress has been made in creating digital atlases for the mouse brain, covering regions, cell composition and connectivity. For the spinal cord, however, such atlases do not yet exist. In this dissertation, I present the first versions of high-resolution 3D atlases of the spinal cords for five mammalian species: mouse, rat, marmoset, rhesus monkey, and human. First, I summarize the current state of the efforts to understand the spinal cord and its organization principles from anatomy (cytoarchitecture) to cell classifications and connectome principles. I then describe how the different properties of the spinal cord can help us in reconstructing the various spinal structures. Next, I present a workflow for data-driven reconstruction that can be used to create 3D models of the spinal cord of any species based on available annotated 2D microscopy images. I used data from Sengul, Watson, et al., 2012 to generate 3D reference volumes of the cytoarchitecture of the 5 mammalian species: mouse, rat, marmoset, rhesus monkey and human. Every 3D reference volume model comprises six levels of the spinal cytoarchitecture and defines their locations, shapes, and volumes. This allows us to compare the spinal cord of these 5 species. I find that all reconstructed spinal cords are remarkably similar, showing far fewer differences than the brains of the respective species. For the mouse spinal cord, I use data from the Allen Institute for Brain Science to generate a cell atlas for the three major spinal cell types: interneurons, motoneurons, and glial cells. The mouse spinal cell atlas comprises the densities and positions of 11,7 million cells in total with 2.7 million interneurons, about 31 thousand motoneurons, and 9.0 million glial cells. Next, I discuss how recent transcriptomic data can be used to assign a gene expression profile to each neuron in the spinal cell atlas. I present a preliminary analysis of RNA-sequencing and in situ RNA hybridization data aiming to develop a workflow for labeling of spinal cells with appropriate transcriptome profiles. Finally, I summarize the results of my dissertation as well as consequences and limitations of the underlying assumptions. This discussion ends with an outlook for how the atlas can be developed further in the future.</p>
<p><img src="bioexplorer_spinal_cord_banner.png" alt="" class="inline" title="___"/></p>
<p>References:</p><ul>
<li><a href="https://infoscience.epfl.ch/record/302106">Digital reconstruction of the mammalian spinal cord: from anatomical reference volume to cell type atlas of the mouse spinal cord</a></li>
</ul>
<h3><a class="anchor" id="autotoc_md21"></a>
Simulation of neuronal activity</h3>
<p>Using the NEURON simulation package, the circuit information is loaded from disk, instantiating the various cell models (morphologies with ion channel distribution) and synaptic connections. The experimenter selects a stimulus protocol which will inject electrical current into the network and increase the membrane voltages of cells. As cells approach a threshold current, they release an action potential (AP) which will then propagate additional current changes to other cells via the synapses' release mechanisms. <em>BBBE</em> loads the simulation reports generated by NEURON and maps the voltages to the corresponding segments of the morphologies. A transfer function defines the mapping between a color and a voltage value.</p>
<h3><a class="anchor" id="autotoc_md22"></a>
Morphology synthesis</h3>
<p>The goal of <a href="https://www.sciencedirect.com/science/article/pii/S2211124722003308">computational synthesis of cortical dendritic morphologies</a> is to be able to generate an arbitrary number of neurons (and also other cells, such as glia) that can be subsequently used in various types of simulation. Part of this goal is to recreate in the synthesized cells as many morphological features as possible.</p>
<p><img src="synthesis.png" alt="" class="inline" title="___"/></p>
<p>The synthesis scheme is based on the assumption that it is necessary to know the environment within which the cells are growing in order to recreate them accurately. Neuronal morphologies are influenced both by the embedding space and the presence of other cells. Their axons may target certain regions or the dendrites may mass in one region to collect input, such as the apical tuft of pyramidal cells. It is important therefore to synthesize the cells within biologically accurate volumes.</p>
<h3><a class="anchor" id="autotoc_md23"></a>
Proximity detection</h3>
<p>In the context of brain simulation, detecting touches between neurons is a essential part of the process. The <em>BBBE</em> provides a renderer that computes the distance between the geometries in the 3D scene.</p>
<p><img src="touchdetection.png" alt="" class="inline" title="___"/></p>
<p>When a ray hits a geometry, a random secondary ray is sent in a direction belonging to an hemisphere defined by the normal to the surface. If that secondary ray hits another geometry, the distance between the initial hit and the new intersection is computed, and the corresponding color is assigned to the pixel. By default, red is for short distances (including touches), and green for longer ones. The notion of short and long is defined in the settings of the renderer.</p>
<h3><a class="anchor" id="autotoc_md24"></a>
White matter and tractography</h3>
<h4><a class="anchor" id="autotoc_md25"></a>
White matter</h4>
<p>White matter is composed of bundles, which connect various grey matter areas (the locations of nerve cell bodies) of the brain to each other, and carry nerve impulses between neurons. Myelin acts as an insulator, which allows electrical signals to jump, rather than coursing through the axon, increasing the speed of transmission of all nerve signals.</p>
<p>A <a href="./bioexplorer/pythonsdk/notebooks/connectomics/BioExplorer_import_white_matter_allen_brain.ipynb">Python notebook example</a> demonstrates how to download and import white matter streamlines from the <a href="https://alleninstitute.org">Allen Brain Institute website</a> into the <em>BBBE</em> database. Another <a href="./bioexplorer/pythonsdk/notebooks/connectomics/BioExplorer_white_matter.ipynb">Python notebook</a> demonstrates how to visualize the streamlines with the <em>BBBE</em>.</p>
<div class="fragment"><div class="line">white_matter_model = be.add_white_matter(</div>
<div class="line">    assembly_name=&#39;White matter&#39;,  population_name=&#39;connectome&#39;, radius=2.5)</div>
</div><!-- fragment --><p><img src="white_matter.png" alt="" class="inline" title="___"/></p>
<p>References:</p><ul>
<li><a href="http://connectivity.brain-map.org/">Allen Brain Institute: mouse connectivity projections</a></li>
<li><a href="https://alleninstitute.github.io/AllenSDK/connectivity.html">Allen Brain Institute: mouse connectivity atlas</a></li>
</ul>
<p>The white matter feature contributed to generating images for the <a href="https://www.nature.com/articles/s41467-019-11630-x">null model of the mouse whole-neocortex micro-connectome</a> publication and produce the corresponding movie.</p>
<div align="center"> <a href="https://www.youtube.com/watch?v=6cBNdAn04Qw"><img src="https://github.com/BlueBrain/BioExplorer/blob/master/bioexplorer/pythonsdk/doc/source/images/connectomics_movie_preview.png" alt="" style="width:50%;" class="inline"/> </a> </div><p><a class="anchor" id="autotoc_md26"></a> </p><h5>Tractography</h5>
<p>Tractography is a neuroimaging technique used in the field of neuroscience and medical imaging to visualize and study the white matter pathways in the human brain. White matter consists of bundles of nerve fibers (axons) that connect different regions of the brain and enable communication between them. Tractography helps researchers and clinicians map and understand these complex neural pathways.</p>
<p>The basic principle of tractography involves tracking the diffusion of water molecules in brain tissue. This technique is often referred to as diffusion-weighted magnetic resonance imaging (DW-MRI) or diffusion tensor imaging (DTI). Here's how it works:</p>
<ol type="1">
<li>Diffusion of Water Molecules: Water molecules in brain tissue naturally diffuse in a preferred direction along the axonal fibers in white matter. This diffusion is hindered by cellular structures, membranes, and myelin sheaths surrounding axons.</li>
<li>Imaging Process: During a DW-MRI scan, multiple images are acquired with different gradients of magnetic fields, which allows for the measurement of water diffusion in multiple directions within each voxel (3D pixel) of the brain.</li>
<li>Data Analysis: The data from DW-MRI is processed to calculate a diffusion tensor, which provides information about the direction and magnitude of water diffusion in each voxel. From this information, researchers can infer the orientation of axonal fibers in that region.</li>
<li>Tractography Reconstruction: Tractography algorithms use the diffusion tensor information to reconstruct and visualize the trajectories of white matter pathways throughout the brain. This results in colorful images and 3D representations that show the connections between different brain regions.</li>
</ol>
<p>Tractography has numerous applications in neuroscience and clinical practice. Researchers use it to study brain connectivity, understand the organization of neural circuits, and investigate neurological disorders. In the clinical setting, tractography can help surgeons plan brain surgeries, evaluate the extent of brain lesions, and assess the impact of brain injuries or diseases on white matter pathways. It has also been used to study conditions like Alzheimer's disease, multiple sclerosis, and stroke.</p>
<p>Different variations of tractography algorithms have been developed to improve the accuracy of tracking neural pathways, such as deterministic tractography and probabilistic tractography. Overall, tractography is a valuable tool for advancing our understanding of brain structure and function.</p>
<p><img src="tractography.png" alt="" class="inline" title="___"/></p>
<p>A <a href="./bioexplorer/pythonsdk/notebooks/tractography/BioExplorer_ISMRM_2015_Tracto_challenge_ground_truth_bundles_TRK_v2.ipynb">Python notebook example</a> demonstrates how to download streamlines from the <a href="https://zenodo.org/record/572345">Tractography Challenge ISMRM 2015</a> dataset and visualize them in the <em>BBBE</em>.</p>
<p>References:</p><ul>
<li><a href="https://zenodo.org/record/572345">Tractography Challenge ISMRM 2015</a></li>
</ul>
<h3><a class="anchor" id="autotoc_md27"></a>
Enzyme reactions</h3>
<p>An enzyme attracts substrates to its active site, catalyzes the chemical reaction by which products are formed, and then allows the products to dissociate (separate from the enzyme surface). The combination formed by an enzyme and its substrates is called the enzyme–substrate complex. The <em>BBBE</em> allows easy visualization of enzyme reactions by providing a substrate, a product, and a type of reaction (for example: <a href="./bioexplorer/pythonsdk/notebooks/molecular_systems/BioExplorer_enzyme_reaction.ipynb">Hexokinase</a>).</p>
<p><img src="bioexplorer_enzyme_reactions_banner.png" alt="" class="inline" title="___"/></p>
<p>References:</p><ul>
<li><a href="https://bbp.epfl.ch/ngv-portal/">Blue Brain Project Neuro-Glia-Vasculature Portal</a></li>
<li><a href="https://bbp.epfl.ch/ngv-portal/#explore">Blue Brain Neuro-Glia-Vasculature Portal. Metabolism. Reconstruction Data</a></li>
<li><a href="https://www.biorxiv.org/content/10.1101/2023.08.30.555341v1">Breakdown and rejuvenation of aging brain energy metabolism</a></li>
</ul>
<h3><a class="anchor" id="autotoc_md28"></a>
Neuromodulation</h3>
<p>Neuromodulation of neocortical microcircuits is one of the most fascinating and mysterious aspects of brain physiology. Despite over a century of research, the neuroscientific community has yet to uncover the fundamental biological organizing principles underlying neuromodulatory release. Phylogenetically, Acetylcholine (ACh) is perhaps the oldest neuromodulator, and one of the most well-studied. ACh regulates the physiology of neurons and synapses, and modulates neural microcircuits to bring about a reconfiguration of global network states. ACh is known to support cognitive processes such as learning and memory, and is involved in the regulation of arousal, attention and sensory processing. While the effects of ACh in the neocortex have been characterized extensively, integrated knowledge of its mechanisms of action is lacking. Furthermore, the ways in which ACh is released from en-passant axons originating in subcortical nuclei are still debatable. Simulation-based paradigms play an important role in testing scientific hypotheses, and provide a useful framework to integrate what is already known and systematically explore previously uncharted territory. Importantly, data-driven computational approaches highlight gaps in current knowledge and guide experimental research. To this end, I developed a multi-scale model of cholinergic innervation of rodent somatosensory cortex comprising two distinct sets of ascending projections implementing either synaptic (ST) or volumetric transmission (VT). The model enables the projection types to be combined in arbitrary proportions, thus permitting investigations of the relative contributions of these two transmission modalities. Using our ACh model, we find that the two modes of cholinergic release act in concert and have powerful desynchronizing effects on microcircuit activity. Furthermore we show that this modeling framework can be extended to other neuromodulators, such as dopamine and serotonin, with minimal constraining data. In summary, our results suggest a more nuanced view of neuromodulation in which multiple modes of transmitter release - ST vs VT - are required to produce synergistic functional effects.</p>
<p><img src="bioexplorer_neuromodulation_banner.png" alt="" class="inline" title="___"/></p>
<p>References:</p><ul>
<li><a href="https://infoscience.epfl.ch/record/294819">Neuromodulation of neocortical microcircuitry: a multi-scale framework to model the effects of cholinergic release</a></li>
<li><a href="https://www.sciencedirect.com/science/article/abs/pii/S030645220100389X">The cholinergic innervation develops early and rapidly in the rat cerebral cortex: a quantitative immunocytochemical study</a></li>
</ul>
<h1><a class="anchor" id="autotoc_md29"></a>
Extensions</h1>
<p>The <em>BBBE</em> represents a highly adaptable and flexible system tailored to explore a wide array of scientific datasets. This exceptional platform has been engineered to accommodate a diverse range of data sources and formats. Its extensibility lies in the incorporation of plugins, which serve as add-on modules expanding the system's capabilities.</p>
<p>Among the plugins available in the extensions folder, there are dedicated modules catering to specific scientific domains. For instance, the cosmology plugin empowers researchers to delve into and analyze astronomical datasets. On the other hand, the medical imaging plugin facilitates in-depth investigations of medical datasets, making it possible for healthcare professionals and scientists to examine and interpret various types of medical images with remarkable precision.</p>
<p>Beyond these examples, <em>BBBE</em> boasts the potential to accommodate additional plugins, making it a powerful platform for scientists from diverse disciplines to explore and make sense of complex scientific data efficiently and effectively.</p>
<h2><a class="anchor" id="autotoc_md30"></a>
Space explorer</h2>
<p>Scientific visualization plays a vital role in the field of space exploration by enabling researchers to interpret and communicate complex astronomical data in a visually compelling manner. It involves the use of advanced computer graphics techniques to transform raw data collected from telescopes, satellites, and spacecraft into visually stunning and informative representations. Through scientific visualization, astronomers can create accurate and detailed models of celestial objects, such as galaxies, nebulae, and planetary systems, aiding in the understanding of their formation, evolution, and behavior. It allows scientists to explore and analyze vast datasets, unveiling patterns, relationships, and hidden structures that may otherwise go unnoticed. Additionally, scientific visualization serves as a powerful tool for public outreach, engaging and inspiring the general public by presenting the awe-inspiring beauty and scientific discoveries of the cosmos in accessible and captivating ways. Overall, scientific visualization plays a crucial role in advancing our knowledge of the universe and fostering a deeper appreciation for the wonders of space exploration.</p>
<h3><a class="anchor" id="autotoc_md31"></a>
Black holes</h3>
<p>The port of <a href="https://www.shadertoy.com">Shadertoy</a>'s black hole implementation to the <em>BBBE</em> represents an exciting fusion of artistic creativity and scientific exploration. Shadertoy, a popular platform for shader development and visual effects, offers a stunning black hole simulation that captures the mesmerizing visual aspects of these cosmic entities. The integration of Shadertoy's black hole into the <em>BBBE</em> allows users to engage with the visualization in a more immersive and scalable way.</p>
<p><img src="./extensions/cosmology/docs/blackhole_banner.png" alt="___" class="inline"/></p>
<p>References:</p><ul>
<li>Shadertoy's <a href="https://www.shadertoy.com/view/tsBXW3">Black hole with accretion disk</a></li>
<li><a href="./extensions/cosmology/notebooks/Gaia_ESA.ipynb">Example notebook</a> for visualizing the GAIA ESA archive dataset</li>
</ul>
<h2><a class="anchor" id="autotoc_md32"></a>
Medical Imaging Explorer</h2>
<p>Medical imaging is a vital component of modern healthcare, enabling healthcare professionals to visualize internal structures and diagnose various medical conditions.</p>
<h3><a class="anchor" id="autotoc_md33"></a>
DICOM</h3>
<p>DICOM stands for "Digital Imaging and Communications in Medicine." It is a widely used standard for the storage, transmission, and sharing of medical images and related information in the healthcare industry. DICOM was created by the National Electrical Manufacturers Association (<a href="https://www.nema.org">NEMA</a>) and the American College of Radiology (<a href="https://www.acr.org">ACR</a>) to ensure interoperability and compatibility among various medical imaging devices and software.</p>
<p>The <em>BBBE</em> DICOM plugin implements the loading of DICOM datasets.</p>
<p><img src="./extensions/medicalimaging/docs/dicom_banner.png" alt="___" class="inline"/></p>
<h2><a class="anchor" id="autotoc_md34"></a>
Environment</h2>
<p>Visualizing environmental data holds immense significance in understanding and addressing complex environmental challenges. By transforming vast datasets into clear, graphical representations, we can grasp the intricate interactions between various environmental factors.</p>
<h3><a class="anchor" id="autotoc_md35"></a>
Geological data</h3>
<p>Visualizing geological data is of paramount importance in the field of geology as it provides a powerful means to comprehend and interpret complex information. By converting raw data into visually appealing representations, geologists gain a deeper understanding of the Earth's composition, structure, and processes. Maps, cross-sections, and 3D models enable researchers to identify patterns, trends, and anomalies that may remain obscured in raw datasets. Visualization aids in making informed decisions related to resource exploration, hazard assessment, and environmental management. Moreover, presenting geological findings visually enhances communication with colleagues, policymakers, and the public, fostering greater awareness and support for geological research and its implications for society. The <em>BBBE</em> serves as an indispensable tool in unraveling the mysteries of our planet and harnessing its resources responsibly.</p>
<ul>
<li><a href="./extensions/environment/notebooks/Charm-50.ipynb">Example notebook</a> for visualizing the <a href="http://infoterre.brgm.fr/">InfoTerre</a> dataset</li>
</ul>
<h2><a class="anchor" id="autotoc_md36"></a>
Physics</h2>
<h3><a class="anchor" id="autotoc_md37"></a>
Nuclear fusion</h3>
<p>The core challenge lies in modeling the behavior of superheated plasma, the fuel for this revolutionary energy source. To simulate this complex phenomenon, the plasma is envisioned as an array of particles, with each particle intricately characterized by its position and direction. The direction vector of each particle not only represented its position but also served as a direct indicator of the particle's charge, a crucial factor in the fusion reaction.</p>
<p><img src="./extensions/physics/nuclearfusion/notebooks/nuclearfusion_banner.png" alt="___" class="inline"/></p>
<ul>
<li><a href="./extensions/physics/nuclearfusion/notebooks/NuclearFusion.ipynb">Example notebook</a></li>
</ul>
<h2><a class="anchor" id="autotoc_md38"></a>
Python SDK</h2>
<p>A simple API if exposed via the <em>BBBE</em> python library. The API allows scientists to easily create and modify assemblies, according the biological parameters. The <em>BBBE</em> programming language is not necessarily reflecting the underlying implementation, but is meant to be as simple as close as possible to the language used by the scientists to describe biological assemblies.</p>
<p>The <em>BBBE</em> Python SDK is available on <a href="https://pypi.org/project/bioexplorer/">pypi</a>.</p>
<p>A large number of examples (as python notebooks) are provided in the <a href="https://github.com/BlueBrain/BioExplorer/tree/master/bioexplorer/pythonsdk/notebooks">notebooks</a> folder.</p>
<h2><a class="anchor" id="autotoc_md39"></a>
Documentation</h2>
<p>See <a href="https://bluebrain.github.io/BioExplorer/">here</a> for detailed documentation of the source code.</p>
<h2><a class="anchor" id="autotoc_md40"></a>
Deployment</h2>
<p><em>BBBE</em> binaries are publicly available as docker images. <em>BBBE</em> is designed to run in distributed mode, and is composed of 3 modules: A <a href="https://hub.docker.com/r/bluebrain/bioexplorer">server</a>, a <a href="https://hub.docker.com/r/bluebrain/bioexplorer-python-sdk">python SDK</a>, and a <a href="https://hub.docker.com/r/bluebrain/bioexplorer-ui">web user interface</a>. This means that there are 3 docker images to be downloaded on run. Those images can of course run on different machines.</p>
<p><b>Note:</b> The <em>BBBE</em> is NOT a web application. It uses web technologies for the user interface but the <em>BBBE</em> is a single session system that requires one back-end instance per user. There is no limitation in the number of user interface instances that can connect to the back-end, but they will all be sharing the same session.</p>
<p>In this example, we will expose the server on port 5000, the python SDK jupyter notebooks on port 5001, and the user interface on port 5002. One is free to change those ports at will.</p>
<h3><a class="anchor" id="autotoc_md41"></a>
Server</h3>
<div class="fragment"><div class="line">docker run -ti --rm -p 5000:8200 bluebrain/bioexplorer</div>
</div><!-- fragment --><h3><a class="anchor" id="autotoc_md42"></a>
Python SDK</h3>
<div class="fragment"><div class="line">docker run -ti --rm -p 5001:8888 bluebrain/bioexplorer-python-sdk</div>
</div><!-- fragment --><p>Note that a local folder can be specified to access personal notebooks:</p>
<div class="fragment"><div class="line">docker run -ti --rm -p 5001:8888 -v &lt;my_notebooks_folder&gt;:/app/BioExplorer/notebooks bluebrain/bioexplorer-python-sdk</div>
</div><!-- fragment --><h3><a class="anchor" id="autotoc_md43"></a>
Web User Interface</h3>
<div class="fragment"><div class="line">docker run -ti --rm -p 5002:8080 bluebrain/bioexplorer-ui</div>
</div><!-- fragment --><h3><a class="anchor" id="autotoc_md44"></a>
PostgreSQL Database</h3>
<p>A PostgreSQL Database can be created as a Docker image and used by the <em>BBBE</em> as a storage component. More information on how to create and populate the database with sample datasets can be found in the dedicated Docker Database documentation.</p>
<p><img src="BBBE_screenshot.png" alt="" class="inline" title="___"/></p>
<h2><a class="anchor" id="autotoc_md45"></a>
Building from Source</h2>
<h3><a class="anchor" id="autotoc_md46"></a>
BioExplorer</h3>
<h4><a class="anchor" id="autotoc_md47"></a>
Compile</h4>
<p>Run the following statements to build the <em>BBBE</em>.</p>
<div class="fragment"><div class="line">git clone --recursive https://github.com/BlueBrain/BioExplorer.git</div>
<div class="line">mkdir build</div>
<div class="line">cd build</div>
<div class="line">CMAKE_PREFIX_PATH=&lt;brayns_installation_folder&gt; cmake .. -DCMAKE_INSTALL_PREFIX=&lt;brayns_installation_folder&gt;</div>
<div class="line">make install</div>
</div><!-- fragment --><h4><a class="anchor" id="autotoc_md48"></a>
Run</h4>
<p>Run the <em>BBBE</em> with the following command to start the backend:</p>
<div class="fragment"><div class="line">service --http-server localhost:5000 --plugin BioExplorer</div>
</div><!-- fragment --><h2><a class="anchor" id="autotoc_md49"></a>
Simple example</h2>
<p>Considering that the <em>BBBE</em> server is running on the local host, on port 5000, the simplest example to visualize a SARS-COV-2 coronavirus is: </p><div class="fragment"><div class="line">from bioexplorer import BioExplorer</div>
<div class="line">be = BioExplorer(&#39;localhost:5000&#39;)</div>
<div class="line">resource_folder = &#39;../../tests/test_files/&#39;</div>
<div class="line">name=&#39;Coronavirus&#39;</div>
<div class="line">be.add_sars_cov_2(name=name, resource_folder=resource_folder)</div>
</div><!-- fragment --><h1><a class="anchor" id="autotoc_md50"></a>
License</h1>
<p><em>BBBE</em> is available to download and use under the GNU General Public License (<a href="https://www.gnu.org/licenses/gpl.html">GPL</a>, or “free software”). The code is open sourced with approval from the open sourcing committee and principal coordinators of the Blue Brain Project in February 2022.</p>
<h1><a class="anchor" id="autotoc_md51"></a>
Contact</h1>
<p>For more information on <em>BBBE</em>, please contact:</p>
<ul>
<li><a href="https://www.epfl.ch/research/domains/bluebrain/blue-brain/people/our-people/operations-division/cyrille-favreau/"><b>Cyrille Favreau</b></a> - Senior Scientific Visualization Engineer - Blue Brain Project (<a href="cyrille.favreau@epfl.ch">cyrille.favreau@epfl.ch</a>)</li>
<li><a href="https://www.epfl.ch/research/domains/bluebrain/blue-brain/people/our-people/simulation-neuroscience-division/emmanuelle-logette/"><b>Emmanuelle Logette</b></a> - Scientific collaborator - Blue Brain Project (<a href="emmanuelle.logette@epfl.ch">emmanuelle.logette@epfl.ch</a>)</li>
<li><a href="https://www.epfl.ch/research/domains/bluebrain/blue-brain/people/our-people/operations-division/evelyne-schmid-osborne/"><b>Evelyne Schmid Osborne</b></a> - Communications - Blue Brain Project (<a href="evelyne.schmidosborne@epfl.ch">evelyne.schmidosborne@epfl.ch</a>)</li>
<li>Co-authors and contributors:<ul>
<li><a href="https://github.com/tribal-tec">Daniel Nachbaur</a></li>
<li><a href="https://github.com/karjonas">Jonas Karlsson</a></li>
<li><a href="https://github.com/rolandjitsu">Andrei-Roland Groza</a></li>
<li><a href="https://github.com/chevtche">Grigori Chevtchenko</a></li>
<li>Raphael Dumusc</li>
<li>Ahmet Bilgili</li>
<li>Juan Bautista Hernando Vieites</li>
<li><a href="https://github.com/ppodhajski">Pawel Jozef Podhajski</a></li>
<li>Jafet Villafranca Diaz</li>
</ul>
</li>
</ul>
<h1><a class="anchor" id="autotoc_md52"></a>
Funding &amp; Acknowledgment</h1>
<p>The development of this software was supported by funding to the Blue Brain Project, a research center of the École polytechnique fédérale de Lausanne (EPFL), from the Swiss government’s ETH Board of the Swiss Federal Institutes of Technology.</p>
<p>COPYRIGHT 2020–2024 Blue Brain Project/EPFL </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
